gflownet/evaluator/base.py

```diff
@@ -0,0 +1,635 @@
+                z_true = rewards.sum()
+                density_true = rewards / z_true
+                with open(gfn.buffer.test_pkl, "wb") as f:
+                    dict_tt["density_true"] = density_true
+                    pickle.dump(dict_tt, f)
+            hist = defaultdict(int)
+            for x in x_sampled:
+                hist[tuple(x)] += 1
+            z_pred = sum([hist[tuple(x)] for x in x_tt]) + 1e-9
+            density_pred = np.array([hist[tuple(x)] / z_pred for x in x_tt])
+            log_density_true = np.log(density_true + 1e-8)
+            log_density_pred = np.log(density_pred + 1e-8)
+        elif gfn.continuous and hasattr(gfn.env, "fit_kde"):
+            batch, _ = gfn.sample_batch(n_forward=self.logger.test.n, train=False)
+            assert batch.is_valid()
+            x_sampled = batch.get_terminating_states()
+            # TODO make it work with conditional env
+            x_sampled = torch2np(gfn.env.states2proxy(x_sampled))
+            x_tt = torch2np(gfn.env.states2proxy(x_tt))
+            kde_pred = gfn.env.fit_kde(
+                x_sampled,
+                kernel=self.logger.test.kde.kernel,
+                bandwidth=self.logger.test.kde.bandwidth,
+            )
+            if "log_density_true" in dict_tt and "kde_true" in dict_tt:
+                log_density_true = dict_tt["log_density_true"]
+                kde_true = dict_tt["kde_true"]
+            else:
+                # Sample from reward via rejection sampling
+                x_from_reward = gfn.env.sample_from_reward(n_samples=self.logger.test.n)
+                x_from_reward = torch2np(gfn.env.states2proxy(x_from_reward))
+                # Fit KDE with samples from reward
+                kde_true = gfn.env.fit_kde(
+                    x_from_reward,
+                    kernel=self.logger.test.kde.kernel,
+                    bandwidth=self.logger.test.kde.bandwidth,
+                )
+                # Estimate true log density using test samples
+                # TODO: this may be specific-ish for the torus or not
+                scores_true = kde_true.score_samples(x_tt)
+                log_density_true = scores_true - logsumexp(scores_true, axis=0)
+                # Add log_density_true and kde_true to pickled test dict
+                with open(gfn.buffer.test_pkl, "wb") as f:
+                    dict_tt["log_density_true"] = log_density_true
+                    dict_tt["kde_true"] = kde_true
+                    pickle.dump(dict_tt, f)
+            # Estimate pred log density using test samples
+            # TODO: this may be specific-ish for the torus or not
+            scores_pred = kde_pred.score_samples(x_tt)
+            log_density_pred = scores_pred - logsumexp(scores_pred, axis=0)
+            density_true = np.exp(log_density_true)
+            density_pred = np.exp(log_density_pred)
+        else:
+            # TODO: refactor
+            env_metrics = gfn.env.test(x_sampled)
```

Comment:
Deprecated?

Reply:
Yes, this will be re-done in a better way.

Reply:
Good thanks